spacy_llm/ty.py

```diff
@@ -85,20 +95,28 @@ def __call__(self, examples: Iterable[Example], **kwargs) -> Dict[str, Any]:
 
 @runtime_checkable
 class LLMTask(Protocol):
-    def generate_prompts(self, docs: Iterable[Doc]) -> Iterable[_PromptType]:
+    def generate_prompts(
+        self, docs: Iterable[Doc], context_length: Optional[int] = None
+    ) -> Iterable[Tuple[Iterable[_PromptType], Iterable[Doc]]]:
```

Comment:
I think this new design makes sense, but I worry about the basic `LLMTask` becoming too complex for users to implement easily. Additionally, any currently existing user task will not work with `spacy-llm` anymore after this, which would be pretty unfortunate.

I wonder whether we could keep the best of both worlds: allow the more simple `LLMTask` as well as a new `ShardableLLMTask` (naming suggestions welcome). Would that be feasible?

Reply:
I'm not a fan of a parallel task structure, but I see where you are coming from. I'll think about it.

Reply:
I guess the "shardable" one would be a sub class? It does offer additional functionality. It just feels wrong that we need to break the existing API to add those features, especially because not all user tasks might be implementing it.

Reply:
So the concerns are about user tasks, right? Our built-in tasks would remain to be all shardable?
That makes sense, I misunderstood at first. Subclassing won't work here due to different function signatures. I think a different `Protocol` is the cleanest way to go.

I'll draft something.

Reply:
Added a suggestion for this in https://github.com/explosion/spacy-llm/pull/332/commits/c104387059c18d0d79f8808ac50577e22b36998a. In a nutshell: we support shardable and non-shardable tasks now. Non-shardable ones look like they did before this PR, shardable ones have the nested prompt/response structure.

I chose to stick with the nested structure for models, as models are less likely to be user-contributed and it's handy from an implementation perspective if they can process the nested data structures. If we really want models to support both variants as well, we can do that, there's no strong reason not to (other than it doesn't add much of a benefit in my opinion).

Reply:
So if a user currently has an `LLMTask` implemented, and they'd update `spacy-llm` with this PR merged as is, would the built-in models still support that old task variant?

Reply:
Yes.